### ISSUE TITLE
[Pendragon 5e] Values greater than 20 are not reported properly

### ISSUE THREAD

**Issue by clanpotshake** [Roll20/roll20-character-sheets#5976]
Currently rolling a test using the Pendragon sheet reports to the chat window the result of a d20 roll and the value of skill/trait/attribute/passion being rolled. This is correct for values of 20 or less.

In Pendragon, if a Skill, Attribute, Trait or Passion is greater than 20, the threshold for success is set at 20 and the excess applies as a modifier to the d20 roll. So for example, with a Sword skill value of 24, the threshold for success is 20, and the d20 roll receives a +4 modifier. 

(n.b. there's a little bit more to this around how critical hits work in Pendragon, but the sheet doesn't appear to support crits at all)

**Comment by clevett** [comment]
It is a limitation in the roll template. It cannot read the result and do math in this way unfortunately. Particularly given that is has to do math in the other direction as well related to modifiers and fumbles/fails.

To handle this API would need to be used but that isn't available to free users.

Soooo ultimately I left the rolls as they are and trust users to do the math.

- Cassie
_Sir Elliot of Pitton_

**Comment by tontione** [comment]
I am currently working on pendragon french character sheet and fixed this with a roll template.
Not merged yet.
Feel free to contact me and re-use my work. :)

**Comment by clevett** [comment]
I'll happily use it. 

This character sheet has translation support on it. You can use Crowdin to translate it to French.

**Comment by tontione** [comment]
Not sur to understand which sheet @clanpotshake is talking about : pendragon 5.2 or pendragon (which is in fact 5.1) ?
On 5.2 sheet, ther are no modifers on rolls, and on the other one it seems to work (but not in french one, waiting for my fix), but without final result (success/failure/critical/fumble) displayed.

Anyway you can take a look on my work (WIP) on this game : 
https://app.roll20.net/join/6471276/n4oiyA

**Comment by clanpotshake** [comment]
The character sheet I filed this issue for is "King Arthur Pendragon v5.2"
The French character sheet there has the right idea, but fails with very large modifiers.

**Comment by tontione** [comment]
can you give me an exemple of fail with large modifiers ?

**Comment by clanpotshake** [comment]
![image](https://user-images.githubusercontent.com/5251258/78042701-63f01100-7340-11ea-8711-169b7470028d.png)
It looks like you've already tested things a bit, but this is an example of your sheet doing incorrect math with a large modifier. The hover text is a little hard to parse, so in general, any modifier of +20 or more essentially guarantees a critical success, while a modifier of -20 or less guarantees a fumble. Any modifier that brings the target number above 20 makes critical successes more likely, while a target number of 0 or lower makes a fumble (critical failure) more likely.

**Comment by tontione** [comment]
Ok, got it.
In fact i even didn't test this case, because in this situation no need to roll, result is garanteed ! ;)
Anyway i will try to fix it. :)

**Comment by Lockbox313** [comment]
I have planned to take a look at this for Cassie's new 5.2 sheet but I have not time to do it yet.

**Comment by Lockbox313** [comment]
This should do it {dice=[[{1d20+@{sword}-20,1d0+20}kl1 [Roll]]]}} {{threshold=[[{@{sword},20}kl1[Threshold]]]}}

**Comment by Lockbox313** [comment]
@clanpotshake  That is the old sheet you are working on

**Comment by Lockbox313** [comment]
This is the new sheet 
<img width="314" alt="pendragon2" src="https://user-images.githubusercontent.com/8994813/79145549-befe1b00-7db8-11ea-8685-f867193ac9fb.png">

**Comment by clanpotshake** [comment]
@Lockbox313 the screenshot in my comment is from @tontione 's French sheet that they said had a fix for this issue. I opened this issue for the sheet in your screenshot.

**Comment by tontione** [comment]
Thanks for your help.
I'm fixed it and added new improvments (on the "old" french sheet), including : 
new religions according to book of knightds and ladies with dynamic bonus calculation and info,  and associated traits underlined, new roll template for damage rolls, horses (wip)
You can take a look at https://app.roll20.net/join/497536/JVenWw
More to come (glory annual gain fields at least). I'm hopping to send a pull request this week or next.

New sheet is best looking but have less dynamical features, like roll modifiers and results display.
I think it's a design option, and it's good for players to have choice. :)

Comments and suggestions are welcome

**Comment by clevett** [comment]
It was just a quick sheet I made for my own games and then decided to share after the fact.

@Lockbox313 Has this been tested for both fumbles & crits? Will it work for numbers modifiers to be less than 20?

**Comment by Lockbox313** [comment]
@clevett Please do not take any of this as any criticism of a your sheet. 

The code I posted was only for handling mods for abilities above 20.  Crits and Fumbles would be handled by the roll template that is lot easier than trying to pack it into the roll macro.

**Comment by Lockbox313** [comment]
This appears to work but I don't have time to do in depth tests at the minute 

 {{dice=[[{1d20+({@{sword}-20,0}kh1),1d0+20}kl1 [Roll]]]}} {{threshold=[[{@{sword},20}kl1[Threshold]]]}}

**Comment by Lockbox313** [comment]
No matter what you do someone will make a complaint about weird formulae  when they cursor over the die roll .

**Comment by clevett** [comment]
No worries. I took this conversation as constructive feedback. :) Its a question I'd already spent time thinking about before publishing the sheet.

The conclusion I came to was that Criticals and Fumbles needs the API to read the result and return a value based on the rules. How would you implement this with the Roll Template?

**Comment by Lockbox313** [comment]
Here is something I put together based on something I did for the old Pendragon sheet


<rolltemplate class="sheet-rolltemplate-Roll">
	<div class="rollHeader flipColor">{{name}}</div>
	<div class="rollHeader flipColor">{{skill}}</div>

	<div class="rollBody">
		
		{{#rollGreater() threshold 19}}
		
			{{#rollGreater() roll 19}}			
				<div class="rollResult">Roll: <b>20</b> <= {{threshold}}</div>
				<div class="rollResult">Result:Critcal</div>						
			{{/rollGreater() roll 19}}

			{{#^rollGreater() roll 19}}			
				<div class="rollResult">Roll: {{roll}}  <= {{threshold}}</div>
				<div class="rollResult">Result:Success</div>						
			{{/^rollGreater() roll 19}}
			
		{{/rollGreater() threshold 19}}

		
		{{#^rollGreater() threshold 19}}
				{{#rollTotal() roll 20}}
					<div class="rollResult">Roll:{{roll}}  <= {{threshold}} </div>
					<div class="rollResult">Result:Fumble</div>							
				{{/rollTotal() roll 20}}

				{{#^rollTotal() roll 20}}
				
					{{#rollTotal() roll threshold}}
						<div class="rollResult">roll: {{roll}}  <= {{threshold}}</div>
						<div class="rollResult">Result:Critical</div>												
					{{/rollTotal() roll threshold}}				
					{{#^rollTotal() roll threshold}}
					
						{{#rollGreater() roll threshold}}
								<div class="rollResult">Roll: {{roll}}  <= {{threshold}}</div>
								<div class="rollResult">Result:Failure</div>
						{{/rollGreater() roll threshold}}
						{{#^rollGreater() roll threshold}}
								<div class="rollResult">Roll: {{roll}}  <= {{threshold}}</div>
								div class="rollResult">Result:Failure</div>
						{{/^rollGreater() roll threshold}}					
					{{/^rollTotal() roll threshold}}
				{{/^rollTotal() roll 20}}		
		{{^rollGreater() threshold 19}}

	</div>
</rolltemplate>

**Comment by tontione** [comment]
My new sheet is juste merged. It display all results (critical & fumble included)  in all cases.
As Lockbox13 i used roll template without API.  I used a custom prompt modifier.
Mine is much more complex, perhaps is it possible to make it much simpler ...
`{{roll= [[1d20cs[[({@{epee}+floor(?{modif|0}),0}>1)*(@{epee}+floor(?{modif|0}))]]cs>[[(40-[[@{epee}]]-1*floor(?{modif|0})+100*(1-{{[[@{epee}]]+floor(?{modif|0}),0}>21})-1)*(1-({{@{epee}+floor(?{modif|0}),0}>39}))+1]]cf[[20+100*{{@{epee}+floor(?{modif|0}),0}>20}]]+[[{{[[@{epee}+floor(?{modif|0})]],0}>21}*([[@{epee}+floor(?{modif|0})]]-20)]]]]}}`

`<rolltemplate class="sheet-rolltemplate-pendragon">
    < table>
        < tr>
            < th colspan="2" class="middle border-bottom">{{perso}}</th>
        < /tr>
        < tr>
            < th colspan="2" class="middle border-bottom">{{carac_type}} {{carac_name}}< /th>
        < /tr>
        < tr>
            < td class="titre">car.+ mod.</td>
            < td class="valeur">{{carac_value}}+({{modif_context}}) = {{carac_modifiee}}</td>
        < /tr>
        < tr>
            < td class="titre">Résultat</td>
            < td class="valeur">
                {{roll}}
                {{#rollWasCrit() roll}}
                <span class="titre green ">Critique !</span>
                {{/rollWasCrit() roll}}
                {{#^rollWasCrit() roll}}
                {{#rollLess() roll carac_modifiee}}
                <span class="titre green">Réussite </span>
                {{/rollLess() roll carac_modifiee}}
                {{#rollTotal() roll 20}}
                <span class="titre red ">Fumble !</span>
                {{/rollTotal() roll 20}}
                {{#^rollTotal() roll 20}}
                {{#^rollLess() roll carac_modifiee}}
                {{#rollGreater() roll seuil}}
                <span class="titre red ">Echec</span>
                {{/rollGreater() roll seuil}}
                {{/^rollLess() roll carac_modifiee}}
                {{/^rollTotal() roll 20}}
                {{/^rollWasCrit() roll}}
            < /td>
        < /tr>
    <  /table>
<  /rolltemplate>
`

**Comment by tontione** [comment]
why ?

**Comment by Lockbox313** [comment]
> No <⁠table> used for layout. As a general standard a <⁠table> element should only be used for tabular data. The <⁠table> shall not be used for layouts. See Designing Character Sheet Layout for alternatives
https://roll20.zendesk.com/hc/en-us/articles/360037773453 under Good Code

.

**Comment by Lockbox313** [comment]
Ok here is a mod for clevett's sheet  that will take into  consideration skills over 20 and those  under 20.  We need away of handling modifiers so currently it works off the value of the attribute.  Roll results which total to more than 20 will show as 20.  Totals that go below zero should be probably capped to 20.  I haven't delved into SCSS yet so  the styling is in the HTML tag.

Here is an invite to the test game https://app.roll20.net/join/6530468/_j5TEw

**Comment by tontione** [comment]
skill -37, roll 7  => fumble instead of simple failure.

**Comment by tontione** [comment]
Huho, i checked the rule and this point differ between french  & 5.2 english version !
I have to check on 5.0 english version ...

**Comment by Lockbox313** [comment]
In 5.0 edition you automatically fail but roll again and fumble on a  20

> Due to negative modifiers, a statistic value may be temporarily reduced to 0. Negative values are never used in Pendragon, even if a modifier would reduce the value to a negative number. Instead, the value is considered to be 0. In such circumstances, the character who wishes to roll against the modified value receives an automatic failure. Players should still roll to see if they fumble (on a roll
> of 20), as normal.

5.1 is the same as 5.2

**Comment by tontione** [comment]
Old rules, at least on my 3.0.

**Comment by Lockbox313** [comment]
I am confused are you using 3rd edition ?

**Comment by tontione** [comment]
The change is from 5.0 to 5.1
i am using a french translation and english 5.2 .
I thought the french edition was a 5.2 one, and didn't know this difference. I checked the rule on the french version.
As my sheet is a french one, i don't know which rule i have to keep ! :)))

**Comment by Lockbox313** [comment]
I added applied clevett's last update  and added roll queries for abilities and damage to account for modifiers and damage from different types of weapons.  I also added a button to the Squire's Age for Squire Rolls.  The roll template does not automatically roll damage that's something I will look into in the future.

clevett's did you compile your updated scss with the current version of Prepos.  When I used the current version it threw this error "Compound selectors may no longer be extended."  Which I tracked it down to this on line 800  "@extend div.sheet-header;. "  I used another compiler and it worked fine.
